Oheger bosch/upgrade logging to log4j2

oheger-bosch opened the pull request:
This PR updates the logging framework from log4j 1.x to log4j2. The logging output should not be affected by this change.

The dependencies to log4j1 have been replaced by corresponding dependencies to log4j2; for some transitive dependencies to log4j1 explicit excludes have been added. While the log methods have not changed in the new version, the way how to obtain a logger is now different; also the logger classes have been moved to another package. Therefore, all the Java classes that make use of a logger had to be (slightly) adapted.

An initial XML-based configuration file for log4j2 has been added. It should be mainly equivalent to the existing configuration. The configuration could be enhanced later, e.g. to make use of advanced features like asynchronous logging. The properties files to configure log4j1 have been removed.

Note that this PR does not upgrade to the most recent version of log4j2. There is, unfortunately, an incompatibility of the Spring boot version in use with log4j2 versions greater than 2.7. More information can be found [here](https://stackoverflow.com/questions/50680649/log4j2-unrecognized-conversion-specifier-xwex-starting-at-position-160-in-co) and in [this ticket](https://github.com/spring-projects/spring-boot/issues/9172).

### Suggest Reviewer
> You can suggest reviewers here with an @mention.
@blaumeiser-at-bosch @neubs-bsi 

### How To Test?
Deploy the changes in a running server and verify the log output.

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

oheger-bosch committed:
chore(logging): Removed unused dependency to logback.

This dependency was defined in the dependency management section of
the top-level pom, but not used by any sub module.

Signed-off-by: Oliver Heger <60134199+oheger-bosch@users.noreply.github.com>

oheger-bosch committed:
chore(logging): Upgraded from log4j 1 to log4j 2.

Changed the imports and the way a logger is obtained. The logging API
itself stayed compatible.

Note that an older version of log4j2 is used because the Spring Boot
version in use has a compatibility issue with log4j2 versions > 2.7.

Signed-off-by: Oliver Heger <60134199+oheger-bosch@users.noreply.github.com>

oheger-bosch committed:
chore(logging): Switched logging configuration to log4j2 format.

Replaced the old log4j.properties file by an equivalent log4j2.xml
configuration file. Removed all other log4j.properties files; they
should not have any effect on the running system.

Signed-off-by: Oliver Heger <60134199+oheger-bosch@users.noreply.github.com>

oheger-bosch committed:
chore(logging): More tweaks of logging dependencies.

Added the log4j2-specific implementation of Slf4j where it is missing,
excluded transitive dependencies to log4j1. Removed an unused import
to log4j2.

Fixed some conflicts after rebasing on master.

Signed-off-by: Oliver Heger <60134199+oheger-bosch@users.noreply.github.com>

oheger-bosch committed:
chore(logging): Updated OSGi package imports.

Replaced imports referencing log4j1 by the corresponding package for
log4j2.

Signed-off-by: Oliver Heger <60134199+oheger-bosch@users.noreply.github.com>

blaumeiser-at-bosch commented on backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java:
```diff
@@ -14,6 +14,7 @@
 import com.google.common.base.Strings;
 import com.google.common.collect.*;
 
+import org.apache.logging.log4j.LogManager;
```
Would it make sense to have the imports sorted and close by each other? Please consider this in all cases.

oheger-bosch replied:
Logging-related imports have been ordered in all files that have been touched (unless I missed some ;-)

blaumeiser-at-bosch commented on backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentSearchHandler.java:
```diff
@@ -9,11 +9,12 @@
  */
 package org.eclipse.sw360.datahandler.db;
 
+import org.apache.logging.log4j.LogManager;
```
Import should be close to the other log import

neubs-bsi commented on backend/utils/src/main/java/org/eclipse/sw360/attachments/db/RemoteAttachmentDownloader.java:
```diff
@@ -50,36 +49,36 @@ public static int retrieveRemoteAttachments(Supplier<HttpClient> httpClient, Str
 
         List<AttachmentContent> remoteAttachments = attachmentContentRepository.getOnlyRemoteAttachments();
-        log.info(format("we have %d remote attachments to retrieve", remoteAttachments.size()));
+        log.info("we have {} remote attachments to retrieve", remoteAttachments.size());
```
maybe remove the "we have". it is kind of weird

oheger-bosch replied:
In the first 2 or 3 files I modified I adapted the log messages, but then noticed that placeholders are typically not used in the codebase. (Maybe this is one reason for the statement that logging slows down performance.) In this PR I would like to focus on the update of the logging dependency, so these changes are easier to review. The proper use of the logging API deserves a PR on its own.

neubs-bsi reviewed (commented):
I noticed that in some files you already changed the log messages to be formatted without a specific formatting function and nicely placing the objects at the end of the method. But in most files the log messages weren't changed. Is that on purpose?

blaumeiser-at-bosch commented on backend/pom.xml:
```diff
@@ -48,9 +48,16 @@
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+        </dependency>
+        <dependency>
```
Where is the slf4j stuff needed?

oheger-bosch replied:
Slf4j is still used by some parts of the codebase, also by the generated Thrift code and in transitive dependencies. This import makes sure that all log statements written through the slf4j API are routed to log4j.

blaumeiser-at-bosch reviewed (changes requested):
Two small things

oheger-bosch committed:
chore(logging): Fixed test failures caused by NoClassDefFound errors.

In projects that depend on spdx-tools, the log4j1.x API classes must
be available. This is now achieved by including the corresponding
artifact from the log4j2 project.

Signed-off-by: Oliver Heger <60134199+oheger-bosch@users.noreply.github.com>

oheger-bosch committed:
chore(logging): Added a library containing the log4j2 classes.

The library produces a bundle that contains the log4j2 jars in use and
exposes their packages.

Tweaked the bnd file for the portlet bundle: the package import for
the old log4j package is still needed by transitive dependencies.

Signed-off-by: Oliver Heger <60134199+oheger-bosch@users.noreply.github.com>
